File: 1-two-sum/swift/solution.swift
// swift solution.swift

class Solution {
    func twoSum(_ nums: [Int], _ target: Int) -> [Int] {
        var dict = [Int: Int]()
        for (index, num) in nums.enumerated() {
            let diff = target - num
            if let preIndex = dict[diff] {
                return [preIndex, index]
            }
            dict[num] = index
        }
		return [-1, -1]
    }
}

func test(_ nums: [Int], _ target: Int, _ indices: [Int]) {
    let s = Solution()
    let ret = s.twoSum(nums, target)
    if ret == indices {
        print("pass: \(nums)")
    } else {
        print("fail: \(nums)")
    }
}

test([2, 7, 11, 15], 9, [0, 1])
test([3, 2, 4], 6, [1, 2])
test([0, 4, 3, 0], 0, [0, 3])
test([-3, 4, 3, 90], 0, [0, 2])
test([-1, -2, -3, -4, -5], -8, [2, 4])

File: 1-two-sum/scala/solution.scala
//
// scala solution.scala
//
object Solution {
    def twoSum(nums: Array[Int], target: Int): Array[Int] = {
        var m: Map[Int,Int] = Map()
        for (index <- 0 until nums.length) {
            val num = nums(index)
            val diff = target - num
            if (m.contains(diff)) {
                return Array(m(diff), index)
            }
            m += (num -> index)
        }
        return Array(-1, -1)
    }

    def test(nums: Array[Int], target: Int, indices: Array[Int]): Unit = {
        val ret = twoSum(nums, target)
        if (ret(0) == indices(0) && ret(1) == indices(1)) {
            println("pass " + nums.mkString("[", ", ", "]"))
        } else {
            println("fail " + nums.mkString("[", ", ", "]"))
        }
    }

    def main(args: Array[String]): Unit = {
        test(Array(2, 7, 11, 15), 9, Array(0, 1))
        test(Array(3, 2, 4), 6, Array(1, 2))
        test(Array(0, 4, 3, 0), 0, Array(0, 3))
        test(Array(-3, 4, 3, 90), 0, Array(0, 2))
        test(Array(-1, -2, -3, -4, -5), -8, Array(2, 4))
    }
}
